tests/src/test/scala/cats/tests/IndexedStateTSuite.scala

```diff
@@ -269,6 +269,41 @@ class IndexedStateTSuite extends CatsSuite {
     result.run(()).value should === (((), ()))
   }
 
+  test("untilDefinedM works") {
+    val counter = State { i: Int =>
+      val res = if (i > 100) Some(i) else None
+      (i + 1, res)
+    }
+
+    counter.untilDefinedM.run(0).value should === ((102, 101))
+
+    OptionT(counter).untilDefined.run(0).value should === ((102, 101))
+  }
+
+  test("foreverM works") {
+    val step = StateT[Either[Int, ?], Int, Unit] { i =>
+      if (i > 1000) Left(i) else Right((i + 1, ()))
+    }
+    step.foreverM.run(0) match {
+      case Left(big) => big should === (1001)
+      case Right((_, _)) => fail("unreachable code due to Nothing, but scalac won't let us match on it")
+    }
+  }
```

Comment:
nitpick. The current implementation looks stack safe, but shall we still add a stack safety test for future guard rail?